util/fetch_deps.py

```diff
@@ -0,0 +1,204 @@
+        base = os.path.dirname(base)
+        if not os.path.isdir(base):
+            return False
+
+        tries -= 1
+        if tries < 1:
+            return False
+
+    return base
+
+
+def safe_rmtree(path):
+    tries = 30
+    while tries > 0:
+        tries -= 1
+
+        try:
+            shutil.rmtree(path)
+            break
+        except Exception:
+            if tries <= 0:
+                raise
+
+            time.sleep(0.3)
+
+
+def main(update_hashes=False):
+    base = find_basedir()
+    scope = {}
+
+    if not base:
+        print('ERROR: tools/DEPS.py not found!')
+        sys.exit(1)
+
+    deps_path = os.path.join(base, "util", 'DEPS.py')
+    with open(deps_path, 'r') as stream:
+        exec(stream.read(), scope)
+
+    deps = scope.get('deps', {})
+    cache = {}
+    cache_path = os.path.join(base, "util", 'DEPS.cache')
+    dl_path = os.path.join(base, "util", '.deps_dl')
+
+    if os.path.isfile(cache_path):
+        with open(cache_path, 'r') as stream:
+            cache = json.load(stream)
+
+    old = set(cache.keys())
+    new = set(deps.keys())
+
+    missing = new - old
+    obsolete = old - new
+    outdated = set()
+
+    for key, meta in deps.items():
+        if not os.path.isdir(os.path.join(base, meta['dest'])):
+            missing.add(key)
+        elif 'hash' in meta and key in old and cache[key].get('hash', (None, None))[1] != meta['hash'][1]:
+            outdated.add(key)
+
+    if os.path.isdir(dl_path):
+        print('Removing left overs...')
+        safe_rmtree(dl_path)
+
+    os.mkdir(dl_path)
+    rep_map = {}
+
+    try:
+        if missing | outdated:
+            print('Fetching missing or outdated dependencies...')
+            count = len(missing | outdated)
+            for i, key in enumerate(missing | outdated):
+                print('[%3d/%3d]: %s' % (i + 1, count, key))
+
+                meta = deps[key]
+                dlname = os.path.join(dl_path, os.path.basename(meta['url']).split('.', 1)[1])
+                link = meta['url'].split('#')[0]
+                dest = os.path.join(base, meta['dest'])
+
+                subprocess.check_call(['curl', '-Lo', dlname, link])
+
+                if 'hash' in meta:
+                    print('Hashing...')
+                    h = hashlib.new(meta['hash'][0])
+
+                    with open(dlname, 'rb') as stream:
+                        h.update(stream.read(16 * 1024))
+
+                    if update_hashes:
+                        rep_map[meta['hash'][1]] = meta['hash'][1] = h.hexdigest()
+
+                    elif h.hexdigest() != meta['hash'][1]:
+                        print(h.hexdigest())
+                        print(meta['hash'][1])
+                        print('ERROR: %s failed the hash check.' % key)
+                        break
+
+                if os.path.isdir(dest):
+                    print('Removing old files...')
+                    safe_rmtree(dest)
```

Comment:
This also deletes the dummy files, unfortunately.  The dummy files existed to create the directories on disk that were needed, as you can't store an empty directory in git.

Maybe it would be better to put the instructions into the markdown file and delete the dummy text files? Alternatively, move it into a markdown file in the ThirdParty directory (which doesn't get deleted).

Reply:
I think the instructions would be easier to find in the README file. The problem with the dummy files is that there's no reason to look at them until you get some kind of error about missing files in that directory.

The script was originally written with the assumption that the directories listed in the `DEPS` file would be completely managed by the script (including updating) and removing the whole directory is the safest way to get rid of old files from previous dependency versions.